Network manager updates - Roadmap Issue #23 updates

alexandercyu opened the pull request:
Completed Items
-------------------
- Admin form page 
- Saving Admin form to database  (Updating database entry)
- Getting information about specific node from database as json 
- Query github node registry json and store in database

alexandercyu committed:
Changed node manager app to network manager app

- Created and initialized database with sample node info.
- Added views to view database info, add new info to database, and edit database

alexandercyu committed:
Renamed folder to 'database'

alexandercyu committed:
Moved main.py

Moved app/main.py to go under database folder

alexandercyu committed:
Moved __init__.py

Moved api/__init__.py to go under database folder

alexandercyu committed:
create __init__.py

Add routes for views

alexandercyu committed:
Move __init__.py

Move home/__init__.py to go under database folder

alexandercyu committed:
Move 'home' files

Move 'home' files to go under database folder

alexandercyu committed:
Move main.py

Move api/main.py to go under database folder

alexandercyu committed:
Update initialize_db.py

Create database functions.

Includes: functions to retrieve node_registry.json from github
              : functions to create database

TODO: add function to add github node_registry records to database

alexandercyu committed:
Create models.py

Create definition for 'nodes' table

alexandercyu committed:
Create node_registry.py

Adds functions to get node_registry.json from github and save the contents under database/settings.json

TODO (To Be Considered):
Function  to reformat incoming json as:
UofT{
node-name: ...
url: ...
}
PAVICS{
node-name:...
url: ...
}

alexandercyu committed:
Create settings.json

Settings file created from github node_registry.json

alexandercyu committed:
Create network manager template files

Template files for network manager web interface.

NOTE: node_register.pt may not be needed, but will be kept in case future needs require it

alexandercyu committed:
Create views.py

Creates views for network manager

alexandercyu committed:
Create setup.py

Sets up pyramid app for network manager

alexandercyu committed:
Added DB update function

Added function to update 'nodes' table with new urls for an existing node name or insert 'nodes' table with a new node name and url entry

alexandercyu committed:
Update initialize_db.py

Cleaned up comments and empty lines

alexandercyu committed:
Update node_registry.py

Added 'app_settings.json' that contains values the app will use (eg. github node registry raw url)

alexandercyu committed:
Create app_settings.json

Settings file for the app

alexandercyu committed:
Update node_info.pt

Now only outputs the node details as a json, as per stated requirements

alexandercyu committed:
Made backup copies of templates

Backed up templates of node_info.pt and node_registry.pt

alexandercyu committed:
Update node_register.pt

Changed node_register.pt template to use Deform form template system

alexandercyu committed:
Update views.py

Changed NodeForm schema to define checkboxes and to fit with Deform guidelines

mishaschwartz commented on .gitignore:
Add some comments to this file so that we know what the purpose of the entries are. See https://github.com/Ouranosinc/Magpie/blob/master/.gitignore for inspiration

mishaschwartz commented on development.ini:
```diff
@@ -1,22 +1,49 @@
 
-# Remove pyramid_debugtoolbar before launch
-pyramid.includes = pyramid_debugtoolbar
+sqlalchemy.url = sqlite:///%(here)s/sqltutorial.sqlite
```
We should be using a postgres database (not sqlite) in production (and development too). For now, make this url configurable so that we can easily update to a postgres database in a different pull request.

fmigneault replied:
From experience, I would suggest instead to make the INI file path configurable, and simply override that via volume mount in production. Otherwise, you will need to decode the INI in code to resolve additional settings/env-vars to plug individual parts of the INI configuration.

alexandercyu replied:
- Database switched to postgres
- Database URI held in variable

mishaschwartz commented on database/__init__.py:
```diff
@@ -0,0 +1,45 @@
+from pyramid.config import Configurator
+
```
I'm not going to comment on coding style in this review but before this is pulled in we should clean up all of the additional blank lines and other spacing oddities.

alexandercyu replied:
Removed unnecessary blank lines

mishaschwartz commented on database/__init__.py:
```diff
@@ -0,0 +1,45 @@
+from .models import DBSession, Base, Node
+
+
+def main(global_config, **settings):
```
global_config is not used here. What is it meant to be used for? Can we make use of it somewhere?

fmigneault replied:
This comes from pyramid. It passes around a configurator instance when loading the app.

mishaschwartz commented on database/__init__.py:
Why is the main module for this project named "database"?

alexandercyu replied:
Renamed to network_manager

mishaschwartz commented on database/home/__init__.py:
```diff
@@ -0,0 +1,33 @@
+
+def main(global_config, **settings):
+
+    # Create a 'settings file' called 'home'
```
What does this mean?

alexandercyu replied:
Removed comment

mishaschwartz commented on database/home/templates/home.pt:
This is nice but I don't think it is needed. The only route that actually needs to render html is the one that displays the form used to fill in the information about the node.

alexandercyu replied:
Removed file

mishaschwartz commented on database/home/templates/node_info.pt:
Same here

alexandercyu replied:
Removed file

mishaschwartz commented on database/home/views.py:
I don't think any of these routes are needed either

alexandercyu replied:
Removed routes

mishaschwartz commented on database/node_registry.py:
```diff
@@ -0,0 +1,48 @@
+        return json_content
+
+    def write_json(self, path, data):
+        f = open(path, 'w')
```
Always use with statements when opening files. This can leave a file open indefinitely

alexandercyu replied:
With statement added

mishaschwartz commented on database/node_registry.py:
```diff
@@ -0,0 +1,48 @@
+        node_registry_file = requests.get(node_registry_url, allow_redirects=True)
+        node_registry_json = json.loads(node_registry_file.text)
+
+        self.write_json(settings_path, node_registry_json)
```
We want to store this information in a database, not in a file

alexandercyu replied:
Tests moved to separate folder

mishaschwartz reviewed (changes requested):
Overall this is a good start. I've left some comments and questions below. I also had some general observations:

- I'm no longer able to install the dev dependencies with `pip install -e . ".[dev]"`, I get a package conflict error. Can you have a look and see if setup.py is properly configured?
- It looks like there is a lot of code here that was left over from a tutorial. This makes it really difficult to see what is intended as part of this app and what is not. Please clean up any unused code/files.
- You have created a lot of template files but we really only need one for now: the one that renders the html form that can be used to display the current node settings and update it.

I'd say the main thing to do first is to clean up the existing code. Please go through and remove anything that isn't necessary to create the four routes that we discussed. After that it will be easier to test out this PR and I can provide some more concrete feedback.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

alexandercyu committed:
Code style fixes applied and project renamed

- renamed database folder (app name) to network_manager
- applied code style fixes such as trimming whitespaces and removing blank lines
- removed unneeded folders such as 'home'

alexandercyu committed:
Code style changes made

Applied style checkers for blank lines

alexandercyu committed:
Update README.md

Added commands to set up and start app

alexandercyu committed:
Update __init__.py

Commented out routes for 'node_info' and 'node_register' because those will be accessible through api

alexandercyu committed:
Delete test_db.py

Moved under 'tests' folder

alexandercyu committed:
Add files to test folder

alexandercyu committed:
Update setup.py

Added cornice as requirement

alexandercyu committed:
Update views.py

- Made 'node_info' page accessible via api

- 'node_register' still needs work

alexandercyu committed:
Update main.py

Fixed spelling

alexandercyu committed:
Update initialize_db.py

Moved import to higher in page

alexandercyu committed:
Update models.py

Added blank space

alexandercyu committed:
Update node_registry.py

Removed unneeded lines

alexandercyu committed:
Update node_home.pt

Removed blank lines

alexandercyu committed:
Update README.md

Added instruction

alexandercyu committed:
Returned 'node_register' view to non-api view

Have 'node_register' form run from routes for now

alexandercyu committed:
Update __init__.py

Removed specified HTTP method from 'node_register'

alexandercyu committed:
Update __init__.py

Added route to show successful add in database

alexandercyu committed:
Update initialize_db.py

- moved drop table query to it's own function
- removed commented code

alexandercyu committed:
Create node_added_view.pt

Added new view to show successful record add in database

alexandercyu committed:
Delete unneeded templates

alexandercyu committed:
Update views.py

- remove commented code
- add function to show successfully added new record to database

alexandercyu committed:
Update development.ini

Added variables for database urls

alexandercyu committed:
Update __init__.py

Made comment more specific

alexandercyu committed:
Deleted 'api' folder

Deleted 'api' folder and files inside

alexandercyu committed:
Delete home.pt

Delete home.pt template

alexandercyu committed:
Update node_added_view.pt

Make title dynamic

alexandercyu committed:
Update node_info.pt

Removed blank lines

alexandercyu committed:
Update node_register.pt

- Made page heading dynamic
- changed page title

alexandercyu committed:
Update views.py

Added page title variable node register page and node successful add page

alexandercyu committed:
Update .gitignore

Added /tests/pycache directory

alexandercyu committed:
Update setup.py

Removed commented lines

alexandercyu committed:
Delete test_db.py

Changed test_db.py filename to tests.py

alexandercyu committed:
Create test_expectation.py

Added test_expectation.py to test expected outcome for node info page

alexandercyu committed:
Create tests.py

Changed test_db.py filename to tests.py

alexandercyu committed:
Delete __init__.py

Delete unneeded file

alexandercyu committed:
alembic nodes table scripts

Created upgrade and downgrade functions for 'nodes' table for alembic

alexandercyu committed:
Update .gitignore

Keep alembic files in repo

alexandercyu committed:
Update alembic.ini

Use postgres connection string

alexandercyu committed:
Update development.ini

Use postgres connection string

alexandercyu committed:
Update __init__.py

Change route so it can prepopulate form fields

alexandercyu committed:
Update .gitignore

Add alembic back into gitignore

alexandercyu committed:
Create alembic files

Default files created by alembic

alexandercyu committed:
Update 621a290666b8_create_nodes_table.py

Old alembic head

alexandercyu committed:
Create 9e4cfcf022bf_.py

Current alembic head

alexandercyu committed:
Update db.py

 - Changed name from initialize_db.py to db.py
- fixed error where it was not putting the information from the github node_registry.json into the database

alexandercyu committed:
Delete initialize_db.py

Renamed to db.py

alexandercyu committed:
Update models.py

Added boolean fields in database for checkboxes form fields

alexandercyu committed:
Update node_added_view.pt

Removed unneeded things

alexandercyu committed:
Update node_home.pt

Removed unneeded things

alexandercyu committed:
Update node_register.pt

Removed unneeded things

alexandercyu committed:
Update views.py

- moved NodeForm class so it's inside node_form function
- changed all references of node_register to node_update
- def node_form -> added Boolean schemas for checkboxes
- def node_update -> changed query to update database from registry form

alexandercyu committed:
Renamed files

Renamed from node_register to node_update

alexandercyu committed:
Update db.py

Changed reference from node_registry to node_update

alexandercyu committed:
Update node_home.pt

Changed href to use route url

alexandercyu committed:
Update __init__.py

Changed node_register to node_update

alexandercyu committed:
Update production.ini

Changed references of database to network_manager

alexandercyu committed:
Update setup.py

Removed console script for initializing database
   -> table creation taken care of by alembic

alexandercyu committed:
Update .gitignore

Added alembic pycache folder

alexandercyu committed:
Update README.md

Changed docs to reflect switch to postgres database

alexandercyu committed:
Update app_settings.json

Added entries for alembic

alexandercyu commented on database/__init__.py:
```diff
@@ -0,0 +1,45 @@
+
+
+    # Add a route / path called node_home at /node
+    # config.add_route("home", "/")
```
Commented code removed

alexandercyu commented on setup.py:
```diff
@@ -1,49 +1,48 @@
 from setuptools import setup
 
+# export VENV=~/Documents/GitHub/node-manager/venv
```
Moved necessary lines to readme

alexandercyu commented on setup.py:
```diff
@@ -1,49 +1,48 @@
+        'console_scripts': [
 
-            'main = config:main'
+            'initialize_db = database.initialize_db:main'
```
Alembic used to manage postgres database

alexandercyu committed:
Update __init__.py

Changed comments

alexandercyu commented on database/__init__.py:
```diff
@@ -0,0 +1,45 @@
+    # config.add_route("home", "/")
+
+
+    # Add a route / path called node_home at /node
```
Comments made more descriptive

alexandercyu commented on database/settings/settings.json:
File removed

alexandercyu commented on database/api/main.py:
File removed

alexandercyu commented on database/home/templates/node_register.pt:
```diff
@@ -0,0 +1,98 @@
+<html lang="en">
+
+<head>
+  <title>WikiPage: Add/Edit</title>
```
Removed title

mishaschwartz commented on alembic.ini:
move these settings to development.ini instead of having a separate alembic.ini file

mishaschwartz commented on network_manager/settings/app_settings.json:
ignore what I said in our discussion. Using this is fine for now

alexandercyu replied:
I moved the app_settings stuff into constants.py.  Getting the settings from constants.py was cleaner than reading from a json.

mishaschwartz commented on network_manager/models.py:
```diff
@@ -0,0 +1,49 @@
+    location = Column(Text)
+    affiliation = Column(Text)
+    url = Column(Text, unique=True)
+    weaver = Column(Boolean, default=False)
```
Remove these as per our discussion

alexandercyu replied:
Weaver, catalog, etc. removed.

mishaschwartz commented on production.ini:
```diff
@@ -0,0 +1,55 @@
+listen = localhost:6543
+
+[alembic]
+alembic -c network_manager/alembic/alembic.ini upgrade head
```
Are you sure this runs the upgrade command? If it does, the path to the alembic.ini should be updated to the correct location

alexandercyu replied:
Line removed

alexandercyu commented on Dockerfile:
```diff
@@ -18,4 +18,4 @@ EXPOSE 5000
 # Set the default command for the container to "flask run"
-CMD ["flask", "run"]
\ No newline at end of file
+CMD ["flask", "run"]
```
Dockerfile removed

alexandercyu commented on network_manager/models.py:
```diff
@@ -0,0 +1,49 @@
+    user_email = Column(Text)
+
+
+class Root:
```
Root class used for building access control list

alexandercyu committed:
Deleted unneeded file

File not used anymore

alexandercyu committed:
Deleted file

alexandercyu committed:
Update __init__.py

Removed unneeded route

alexandercyu committed:
Update db.py

-  Deleted unneeded functions
- Changed 'capabilities' into separated 'data' and 'compute' columns
- Renamed  'user_email' into 'support_contact_email'

alexandercyu committed:
Update views.py

- Updated node form definition to include:
       : calendar widget for entering date
       : changed 'user_email' to 'contact_support_email'
       : removed unneeded field definitions
       : Added 'data' and 'compute' fields
       : Removed asterisk indicating required field
       : Removed unneeded views
       : Changed 'node_info' view to render in json directly instead of using template to display json

alexandercyu committed:
Create constants.py

Created constants.py to hold urls and other values that won't change

alexandercyu committed:
Update models.py

 - Removed unneeded column definitions
 - Added other needed column definitions

alexandercyu committed:
Create node_registry.py

- Renamed node_update.py to node_registry.py
- removed unneeded functions (reading json, etc.)

alexandercyu committed:
Update node_update.pt

Added tentative new section to display all available nodes instead of having separate home page display all nodes

alexandercyu committed:
Update production.ini

Removed alembic section

alexandercyu committed:
Update setup.py

Removed debugtoolbar as a requirement

alexandercyu committed:
Added alembic.ini contents to dev and prod ini files

Added alembic section to dev and prod ini files

alexandercyu committed:
Removed Root class

Removed Root class for building access control list since it will be handled by twitcher and magpie

alexandercyu committed:
Update views.py

Remove separate class for date widget

alexandercyu committed:
Update node_registry.py

Add function to get node_info.json

alexandercyu committed:
Update README.md

Update docs for network manager

alexandercyu committed:
Delete alembic.ini

alexandercyu committed:
Update production.ini

Removed debugtoolbar

mishaschwartz commented on tests/tests.py:
```diff
@@ -0,0 +1,81 @@
+import unittest
```
Use pytest or unittest for testing, not both please. We need to keep our tests consistent for maintainability. 

I have a strong preference for using pytest over unittest (if you disagree please let me know and we can discuss).

mishaschwartz commented on tests/test_expectation.py:
please rename these test files so that they are descriptive

alexandercyu commented on network_manager/db.py:
```diff
@@ -0,0 +1,90 @@
+
+
+class DB:
+    def usage(self, argv):
```
Testing methods removed

alexandercyu commented on network_manager/settings/app_settings.json:
```diff
@@ -0,0 +1,5 @@
+{
+  "node_registry_url": "https://raw.githubusercontent.com/DACCS-Climate/DACCS-node-registry/main/node_registry.json",
+  "alembic_ini_file": "../alembic.ini",
```
Removed settings.json

alexandercyu commented on network_manager/alembic/versions/2668c20f914e_create_nodes_table.py:
Unneeded alembic migration files removed

alexandercyu commented on network_manager/node_registry.py:
```diff
@@ -0,0 +1,15 @@
+
+        node_registry_url = NM_NODE_REGISTRY_URL
+        node_registry_file = requests.get(node_registry_url, allow_redirects=True)
+        node_registry_json = json.loads(node_registry_file.text)
```
Replaced json.loads with request json loader

alexandercyu committed:
Update node_registry.py

Use request json loader instead of json package loader

alexandercyu committed:
Change sqlalchemy database url strings

- remove sqlite connection strings
- change alembic section's database connection strings to postres

alexandercyu committed:
Update constants.py

Remove sqlite connection string

alexandercyu committed:
Update .gitignore

Removed sqlite line because the file was removed

alexandercyu committed:
Delete node_home.pt

Removed view template

alexandercyu committed:
Update node_update.pt

Added section to view all nodes at top of page

This may be removed once login accounts are associated with node ids

alexandercyu committed:
Update requirements.txt

Removed unneeded requirements

alexandercyu committed:
Update views.py

- Added blank defaults for loading empty form fields
- Changed node_home() to display_all_nodes() for displaying a blank form with list of all nodes for when user is visiting page for the first time

alexandercyu committed:
Update __init__.py

- Changed node_home route to node_all
- changed path from /node to /node/update with no passed node_id for when user is visiting page for first time

alexandercyu committed:
Update db.py

Added lines for future functionality to get node_info.json

alexandercyu committed:
Update README.md

Updated docs to include production.ini

alexandercyu commented on network_manager/views.py:
```diff
@@ -0,0 +1,162 @@
+        return self.node_form.get_widget_resources()
+
+    @view_config(route_name="node_home", renderer="templates/node_home.pt")
+    def node_home(self):
```
View removed

alexandercyu committed:
Update __init__.py

Updated comments

alexandercyu committed:
Update README.md

Updated docs

alexandercyu commented on README.md:
```diff
@@ -20,3 +20,86 @@ pre-commit install      # Install the pre-commit hooks defined in .pre-commit-co
+
+### For sqlite database
+
+> $VENV/bin/initialize_db development.ini
```
Yes, that's right.  Alembic takes care of the table creation.  I think postgres has to be started separately from this app.  The database 'noderegistry' needs to be created separately as well before starting this app.

mishaschwartz replied:
That makes sense. But... postgres/sqlite won't create the database automatically. Alembic can only create the tables in a database that already exists. What are the instructions to the user to create the database now (before migrating with alembic)?

alexandercyu replied:
Added section on installing Postgres database and setting up the app's database

alexandercyu commented on tests/test_expectation.py:
```diff
@@ -0,0 +1,56 @@
+        )
+        from network_manager.initialize_db import table_exists
+
+        argv = sys.argv
```
This was when sqlite was used as the database and I think it could not create the engine without the development.ini filename being passed to it.

alexandercyu replied:
No, I think when I ran this test I got an error saying the engine could not be created and it was expecting an argument (I think).  Since in earlier code the engine was created through initializing by commandline I thought  the commandline was what was needed.

mishaschwartz replied:
Let's discuss this in person/zoom please

alexandercyu committed:
Update README.md

- Added section on creating postgres database and setting up usernane and password

- Changed section title sizes as necessary

alexandercyu committed:
Update views.py

Changed return for node_info_view to json type instead of dict type

alexandercyu commented on network_manager/views.py:
```diff
@@ -0,0 +1,162 @@
+
+        node_info_json = json.dumps(node_info_dict)
+
+        return dict(node_json=node_info_json)
```
Changed the return

alexandercyu commented on network_manager/views.py:
```diff
@@ -0,0 +1,162 @@
+
+        # Checks for the submit button, then uses POST to get the submitted information
+        # If not, display form using GET
+        if "submit" in self.request.params:
```
Changed request method to POST

alexandercyu committed:
Update views.py

Changed method to POST

alexandercyu committed:
Delete requirements.txt

requirements.txt no longer needed because dependencies are in setup.py

alexandercyu commented on requirements.txt:
```diff
@@ -1,5 +1,8 @@
 flask
```
Deleted requirements.txt

Dependencies already in setup.py

alexandercyu commented on .gitignore:
```diff
@@ -1 +1,17 @@
+
+# App system files
+/network_manager.egg-info/*
+sqltutorial.sqlite
```
Deleted sqlite file

alexandercyu commented on setup.py:
```diff
@@ -1,49 +1,37 @@
+
 # List of dependencies installed via `pip install -e .`
 # by virtue of the Setuptools `install_requires` value below.
 requires = [
```
Removed requirements.txt

alexandercyu commented on network_manager/templates/node_home.pt:
Template removed

alexandercyu committed:
Update .gitignore

Replaced ignore pycache folder lines with single pycache ignore

alexandercyu commented on .gitignore:
```diff
@@ -1 +1,17 @@
+
+# Pycharm system files
+.idea/*
+/tests/__pycache__/*
```
Replaced ignore pycache folder lines with single ignore line

alexandercyu committed:
Update .gitignore

Changed ignore level to ignore directory

alexandercyu commented on .gitignore:
```diff
@@ -1 +1,17 @@
+.DS_Store
+
+# Pycharm system files
+.idea/*
```
Changed ignore to directory level

alexandercyu committed:
Added docstrings

Added docstrings to all functions and classes

alexandercyu committed:
Create network_manager.sqlite

sqlite file created to give user option of using sqlite as database instead of postgres

alexandercyu committed:
Delete test_expectation.py

Renamed test_expectation.py to old test_expectation.py

alexandercyu committed:
Create test_views.py

Added tests for views in views.py

alexandercyu committed:
Create old_test_expectation.py

Renamed test_expectation.py to old_test_expectation.py

alexandercyu committed:
Create pytest.ini and testing.ini

Create ini files for testing

alexandercyu committed:
Create conftest.py

Create testing fixtures

alexandercyu committed:
Create testing.ini

Added testing.ini under 'tests' folder because pytest could not find it

alexandercyu committed:
Update setup.py

Added pytest as testing requirement

alexandercyu committed:
Merged in changes from #11

alexandercyu committed:
Update test_views.py

Added tests for views

alexandercyu committed:
Update test_views.py

- Complete tests for outstanding functions
- exception is Test_node_edit_POST

alexandercyu committed:
Update 242f91dc05ae_create_nodes_table.py

Removed Unique constraint for 'local' column.  As a Boolean column this would cause an error if there were more than one True or False entry

alexandercyu committed:
Update models.py

Removed Unique constraint for 'local' column.  As a Boolean column this would cause an error if there were more than one True or False entry
